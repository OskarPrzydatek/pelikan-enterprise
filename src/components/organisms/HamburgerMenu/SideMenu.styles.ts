import styled, { css } from 'styled-components';

export const SideMenu = styled.div`
  height: 100%;
  width: 100%;
  background-color: ${({ theme }) => theme.colors.white};
  display: grid;
  grid-template-rows: 1fr 20fr 1fr;
  padding: ${({ theme }) => theme.spaces.m};

  @media ${({ theme }) => theme.breakpoints.phone} {
    width: 55%;
  }

  @media ${({ theme }) => theme.breakpoints.tablet} {
    width: 30%;
  }

  @media ${({ theme }) => theme.breakpoints.desktop} {
    width: 20%;
  }
`;

export const Header = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const Nav = styled.nav``;

export const List = styled.ul``;

export const NavItem = styled.li`
  display: flex;
  flex-direction: column;

  button {
    margin: ${({ theme }) => theme.spaces.xs} 0;
    margin-left: ${({ theme }) => theme.spaces.xl};
    font-weight: 700;
    cursor: pointer;
  }
`;

export const currentPageButtonCSS = css`
  color: ${({ theme }) => theme.colors.lightBlue};
`;

export const pageButtonCSS = css`
  color: ${({ theme }) => theme.colors.darkBlue};
`;

export const logoutButtonCSS = css`
  color: ${({ theme }) => theme.colors.white};
`;
