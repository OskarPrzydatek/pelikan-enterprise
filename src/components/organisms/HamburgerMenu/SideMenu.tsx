import React from 'react';

import { Background, Button, Header, Icon } from '~/components/atoms';
import { Icons, Slugs } from '~/constants';
import { sideMenuNavigationItemsMetadata } from '~/metadata';
import { IComponent, ISideMenuNavigationList } from '~/models';

import * as S from './SideMenu.styles';

interface ISideMenuNavigationItem extends ISideMenuNavigationList {
  handleIsCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
}

const SideMenuNavigationItem: React.FC<ISideMenuNavigationItem> = ({
  title,
  items,
  handleIsCurrentPage,
  onNavigate,
}: ISideMenuNavigationItem) => {
  const handleIsCurrentPageStyle = (path: Slugs) =>
    handleIsCurrentPage(path) ? S.currentPageButtonCSS : S.pageButtonCSS;

  return (
    <S.NavItem>
      <Header>{title}</Header>
      {items.map(({ path, label }) => (
        <Button
          key={path}
          css={handleIsCurrentPageStyle(path)}
          dataTestID={`side-menu-item-${path}`}
          variant="ghost"
          onClick={() => onNavigate(path)}
        >
          {label}
        </Button>
      ))}
    </S.NavItem>
  );
};

interface ISideMenu extends IComponent {
  onClickCloseSideMenu: () => void;
  handleIsCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
  onClickLogout: () => void;
}

export const SideMenu: React.FC<ISideMenu> = ({
  onClickCloseSideMenu,
  handleIsCurrentPage,
  onNavigate,
  onClickLogout,
}: ISideMenu) => {
  return (
    <Background onClickBackground={onClickCloseSideMenu}>
      <S.SideMenu data-testid="hamburger-menu-side-menu">
        <S.HeaderBar>
          <Header>Pelikan Enterprise</Header>
          <Icon
            dataTestID="side-menu-close-icon"
            icon={Icons.CLOSE}
            width={15}
            onClick={onClickCloseSideMenu}
          />
        </S.HeaderBar>
        <S.Nav>
          <S.List>
            {sideMenuNavigationItemsMetadata.map(({ title, items }) => (
              <SideMenuNavigationItem
                key={title}
                handleIsCurrentPage={handleIsCurrentPage}
                items={items}
                title={title}
                onNavigate={onNavigate}
              />
            ))}
          </S.List>
        </S.Nav>
        <Button
          css={S.logoutButtonCSS}
          dataTestID="side-menu-logout-button"
          onClick={onClickLogout}
        >
          Wyloguj
        </Button>
      </S.SideMenu>
    </Background>
  );
};
