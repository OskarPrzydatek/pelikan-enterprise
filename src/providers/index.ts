export { StyledComponentsProvider } from './StyledComponents.provider';
