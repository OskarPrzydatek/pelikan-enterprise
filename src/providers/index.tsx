import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { GlobalStyleLayout } from '../styles/layouts';
import { theme } from '../styles/theme';

interface IStyledComponentsProvider {
  children: React.ReactNode;
}

interface ITestingPageProvider {
  element: JSX.Element;
}

export const StyledComponentsProvider: React.FC<IStyledComponentsProvider> = ({
  children,
}: IStyledComponentsProvider) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyleLayout />
      {children}
    </ThemeProvider>
  );
};

export const TestingPageProvider: React.FC<ITestingPageProvider> = ({
  element,
}: ITestingPageProvider) => {
  return (
    <StyledComponentsProvider>
      <MemoryRouter initialEntries={['/page1']}>
        <Routes>
          <Route element={element} path="/:page" />
        </Routes>
      </MemoryRouter>
    </StyledComponentsProvider>
  );
};
