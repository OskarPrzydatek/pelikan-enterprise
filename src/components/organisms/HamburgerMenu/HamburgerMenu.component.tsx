import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { useTheme } from 'styled-components';

import { Icon, IconEnum } from '~/components/atoms';
import { Paths } from '~/router';

import { SideMenu } from './SideMenu.component';

export const HamburgerMenu: React.FC = () => {
  const navigate = useNavigate();
  const theme = useTheme();
  const { pathname } = useLocation();

  const [isSideMenuOpen, setIsSideMenuOpen] = React.useState<boolean>(false);

  const currentPageButtonStyle = (path: Paths) => ({
    color:
      pathname === `/${path}` ? theme.colors.lightBlue : theme.colors.black,
  });

  const onClickOpenSideMenu = () => {
    setIsSideMenuOpen(true);
  };

  const onClickCloseSideMenu = () => {
    setIsSideMenuOpen(false);
  };

  const onNavigate = (path: Paths) => {
    onClickCloseSideMenu();
    navigate(path);
  };

  // TODO: Implement when auth be ready on BE
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const onClickLogout = () => {};

  return (
    <>
      <Icon
        dataTestID="side-menu-hamburger-icon"
        height={15}
        icon={IconEnum.HAMBURGER}
        width={30}
        onClick={onClickOpenSideMenu}
      />
      {isSideMenuOpen && (
        <SideMenu
          currentPageButtonStyle={currentPageButtonStyle}
          onClickCloseSideMenu={onClickCloseSideMenu}
          onClickLogout={onClickLogout}
          onNavigate={onNavigate}
        />
      )}
    </>
  );
};
