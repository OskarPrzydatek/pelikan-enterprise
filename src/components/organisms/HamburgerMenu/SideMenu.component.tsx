import React from 'react';

import {
  Background,
  Button,
  Header,
  Icon,
  IconEnum,
  Text,
} from '~/components/atoms';
import { Slugs } from '~/constants';
import {
  ISideMenuNavigationSection,
  sideMenuNavigationItemsMetadata,
} from '~/metadata';

import * as S from './SideMenu.styles';

interface ISideMenuNavigationItem extends ISideMenuNavigationSection {
  isCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
}

interface ISideMenu {
  onClickCloseSideMenu: () => void;
  isCurrentPage: (path: Slugs) => boolean;
  onNavigate: (path: Slugs) => void;
  onClickLogout: () => void;
}

const SideMenuNavigationItem: React.FC<ISideMenuNavigationItem> = ({
  title,
  items,
  isCurrentPage,
  onNavigate,
}: ISideMenuNavigationItem) => {
  return (
    <S.NavItem>
      <Header>{title}</Header>
      {items.map(({ path, label }) => {
        const handleIsCurrentPageCSS = isCurrentPage(path)
          ? S.currentPageButtonCSS
          : S.pageButtonCSS;

        return (
          <Button
            key={path}
            css={handleIsCurrentPageCSS}
            dataTestID={`side-menu-item-${path}`}
            variant="ghost"
            onClick={() => onNavigate(path)}
          >
            {label}
          </Button>
        );
      })}
    </S.NavItem>
  );
};

export const SideMenu: React.FC<ISideMenu> = ({
  onClickCloseSideMenu,
  isCurrentPage,
  onNavigate,
  onClickLogout,
}: ISideMenu) => {
  return (
    <Background onClickBackground={onClickCloseSideMenu}>
      <S.SideMenu data-testid="hamburger-manu-side-menu">
        <S.Header>
          <Header>Pelikan Business</Header>
          <Icon
            dataTestID="side-menu-close-icon"
            icon={IconEnum.CLOSE}
            width={15}
            onClick={onClickCloseSideMenu}
          />
        </S.Header>
        <S.Nav>
          <S.List>
            {sideMenuNavigationItemsMetadata.map(({ title, items }) => (
              <SideMenuNavigationItem
                key={title}
                isCurrentPage={isCurrentPage}
                items={items}
                title={title}
                onNavigate={onNavigate}
              />
            ))}
          </S.List>
        </S.Nav>
        <Button onClick={onClickLogout}>
          <Text css={S.logoutButtonCSS}>Wyloguj</Text>
        </Button>
      </S.SideMenu>
    </Background>
  );
};
