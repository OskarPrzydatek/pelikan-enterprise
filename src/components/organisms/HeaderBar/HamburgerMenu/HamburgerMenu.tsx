import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

import { Icon } from '~/components/atoms';
import { Icons, Slugs } from '~/constants';
import { IComponent } from '~/models';

import { SideMenu } from '../SideMenu/SideMenu';

interface IHamburgerMenu extends IComponent {}

export const HamburgerMenu: React.FC<IHamburgerMenu> = () => {
  const navigate = useNavigate();
  const { pathname } = useLocation();

  const [isSideMenuOpen, setIsSideMenuOpen] = React.useState<boolean>(false);

  const handleIsCurrentPage = (path: Slugs) => pathname === `/${path}`;

  const onClickOpenSideMenu = () => {
    setIsSideMenuOpen(true);
  };

  const onClickCloseSideMenu = () => {
    setIsSideMenuOpen(false);
  };

  const onNavigate = (path: Slugs) => {
    onClickCloseSideMenu();
    navigate(path);
  };

  // TODO: Implement when auth be ready on BE
  const onClickLogout = () => {};

  return (
    <>
      <Icon
        dataTestID="side-menu-hamburger-icon"
        height={15}
        icon={Icons.HAMBURGER}
        width={30}
        onClick={onClickOpenSideMenu}
      />
      {isSideMenuOpen && (
        <SideMenu
          handleIsCurrentPage={handleIsCurrentPage}
          onClickCloseSideMenu={onClickCloseSideMenu}
          onClickLogout={onClickLogout}
          onNavigate={onNavigate}
        />
      )}
    </>
  );
};
