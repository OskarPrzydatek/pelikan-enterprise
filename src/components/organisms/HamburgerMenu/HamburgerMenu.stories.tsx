import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import { StyledComponentsProvider } from '~/styles';

import { HamburgerMenu } from './HamburgerMenu.component';

const meta: Meta<typeof HamburgerMenu> = {
  title: 'Organisms/HamburgerMenu',
  component: HamburgerMenu,
};

export default meta;
type Story = StoryObj<typeof HamburgerMenu>;

export const HamburgerMenuStory: Story = {
  render: () => (
    <BrowserRouter>
      <StyledComponentsProvider>
        <HamburgerMenu />
      </StyledComponentsProvider>
    </BrowserRouter>
  ),
};
