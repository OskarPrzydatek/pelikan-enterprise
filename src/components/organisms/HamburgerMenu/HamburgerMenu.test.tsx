import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { describe, test, expect } from 'vitest';

import { StyledComponentsProvider } from '~/providers';

import { HamburgerMenu } from './HamburgerMenu.component';

describe('Button', () => {
  test('component snapshot', () => {
    const view = render(
      <BrowserRouter>
        <StyledComponentsProvider>
          <HamburgerMenu />
        </StyledComponentsProvider>
      </BrowserRouter>
    );
    expect(view).toMatchSnapshot();
  });

  test('ensure side menu opening and closing correctly', () => {
    render(
      <BrowserRouter>
        <StyledComponentsProvider>
          <HamburgerMenu />
        </StyledComponentsProvider>
      </BrowserRouter>
    );
    fireEvent.click(screen.getByTestId('side-menu-hamburger-icon'));
    expect(screen.getByTestId('hamburger-menu-side-menu')).toBeInTheDocument();
    fireEvent.click(screen.getByTestId('side-menu-close-icon'));
    expect(
      screen.queryByTestId('hamburger-menu-side-menu')
    ).not.toBeInTheDocument();
  });

  test('ensure side menu navigate to correct page', () => {
    render(
      <BrowserRouter>
        <StyledComponentsProvider>
          <HamburgerMenu />
        </StyledComponentsProvider>
      </BrowserRouter>
    );
    fireEvent.click(screen.getByTestId('side-menu-hamburger-icon'));
    fireEvent.click(screen.getByTestId('side-menu-item-create-offer'));
    expect(
      screen.queryByTestId('hamburger-menu-side-menu')
    ).not.toBeInTheDocument();
    fireEvent.click(screen.getByTestId('side-menu-hamburger-icon'));
    expect(screen.getByTestId('side-menu-item-create-offer')).toHaveStyle(
      'color: #98BFF2'
    );
  });
});
