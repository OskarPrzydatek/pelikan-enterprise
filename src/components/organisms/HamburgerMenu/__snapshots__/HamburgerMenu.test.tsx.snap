// Vitest Snapshot v1

exports[`Button > component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="sc-jSUZER kEKcxS"
        data-testid="side-menu-hamburger-icon"
      >
        <svg
          data-testid="hamburger-icon"
          fill="none"
          height="15"
          viewBox="0 0 45 23"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            stroke="#013059"
            stroke-width="3"
            x2="45"
            y1="21.5"
            y2="21.5"
          />
          <line
            stroke="#013059"
            stroke-width="3"
            x2="45"
            y1="11.5"
            y2="11.5"
          />
          <line
            stroke="#013059"
            stroke-width="3"
            x2="45"
            y1="1.5"
            y2="1.5"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="sc-jSUZER kEKcxS"
      data-testid="side-menu-hamburger-icon"
    >
      <svg
        data-testid="hamburger-icon"
        fill="none"
        height="15"
        viewBox="0 0 45 23"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          stroke="#013059"
          stroke-width="3"
          x2="45"
          y1="21.5"
          y2="21.5"
        />
        <line
          stroke="#013059"
          stroke-width="3"
          x2="45"
          y1="11.5"
          y2="11.5"
        />
        <line
          stroke="#013059"
          stroke-width="3"
          x2="45"
          y1="1.5"
          y2="1.5"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
